RFC,RFT: Switch buildroot to Python 3

ynezz opened the pull request:
This PR switches Python used in buildroot to Python 3 as Python 2.x series [is going to be EOL](https://pythonclock.org/) relatively [soon](https://www.python.org/dev/peps/pep-0373/):

> Being the last of the Python 2.x series, 2.7 will have an extended period of maintenance. Specifically, 2.7 will receive bugfix support until January 1, 2020. After the last release, 2.7 will receive no support

Apart this EOL reason, it seems like Python 3 is going to be needed in order to support Meson/Ninja build system which has gained some traction recently.

* build tested on `Debian GNU/Linux 9.8 (stretch)` with `Python 3.5.3`.
* build tested with ath79 and x86 targets with `CONFIG_ALL=y`, `CONFIG_ALL_NONSHARED=y` and
`CONFIG_ALL_KMODS=y` options.
* run tested on ath79 (Archer C7 v5) and x86_64 (QEMU and apu2).

Please note, that update of scons to 3.0.4 (which supports Python 3) breaks packages from packages feed which build depends on scons. None of them build anymore as it seems, that the SConscripts in those packages are written for Python2.

 * packages/net/iotivity (doesn't build even with latest 1.3.1 release)

 (doesn't build even with latest 1.3.1 release)
```python
   SyntaxError: invalid syntax
    File "/openwrt/build_dir/target-mips_24kc_musl/iotivity-1.2.1/build_common/SConscript", line 40
      print "\nError: Current system (%s) isn't supported\n" % host
```
```python
   LookupError: unknown encoding: string_escape:
    File "/openwrt/build_dir/target-mips_24kc_musl/iotivity-1.3.1/SConstruct", line 28:
      SConscript('build_common/SConscript')
```
 * packages/net/smartsnmpd (seems dead, no commit since 2015):
```python
   SyntaxError: Missing parentheses in call to 'print'
    File "/openwrt/build_dir/target-mips_24kc_musl/smartsnmpd-2014-08-13/SConstruct", line 156
      print "Can't find liblua or liblua5.1!"
```
 * packages/utils/gpsd (doesn't build even with latest 3.18.1 release):
```python
   AttributeError: 'list' object has no attribute 'keys':
     File "/openwrt/build_dir/target-mips_24kc_musl/gpsd-3.18.1/SConstruct", line 1758:
       all_manpages = list(base_manpages.keys()) + list(python_manpages.keys())
```

This PR depends on #1521 as current wireless-regdb version won't build with Python 3. I've included the `wireless-regdb: update to 2019.03.01` commit from #1521 for easier testing of this PR.

### Working targets (build test only with config.seed as used by buildbots for snapshots)

* ath79/generic
* ipq40xx/generic
* lantiq/xway
* x86/64

### Broken targets

* layerscape/armv8_64b

### Broken packages

* package/firmware/layerscape/ls-rcw
* feeds/packages/net/iotivity
* feeds/packages/net/smartsnmpd
* feeds/packages/utils/gpsd

ynezz committed:
wireless-regdb: prefer python provided by make variable

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
scripts/dl_github_archive.py: convert to Python 3 with 2-to-3

Let's convert the script to Python 3.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
scripts/jungo-image: convert to Python 3 with 2-to-3

Let's convert the script to Python 3.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
tools/b43-tools/b43-fwsquash: convert to Python 3 with 2-to-3

Let's convert the script to Python 3.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
build: switch to Python 3

This patch switches Python used in buildroot to Python 3 as Python 2.x
series is going to be EOL relatively soon[1].

 "Being the last of the Python 2.x series, 2.7 will have an extended
  period of maintenance. Specifically, 2.7 will receive bugfix support
  until January 1, 2020. After the last release, 2.7 will receive no
  support."

It seems like Python 3 is going to be needed in order to support Meson
build system which has gained some traction recently.

Build tested on ath79 and x86 with CONFIG_ALL=y, CONFIG_ALL_NONSHARED=y,
CONFIG_ALL_KMODS=y.

Run tested on ath79 (Archer C7 v5) and x86_64 (QEMU and apu2).

1. https://www.python.org/dev/peps/pep-0373/

Ref: PR#1937
Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
tools/scons: switch to Python 3

Build tested on ath79 with following packages from packages feed which
build depends on scons. None of them build anymore as it seems, that the
SConscripts are written for Python2.

 * packages/net/iotivity (KO, doesn't build even with latest 1.3.1 release)

   SyntaxError: invalid syntax
    File "/openwrt/build_dir/target-mips_24kc_musl/iotivity-1.2.1/build_common/SConscript", line 40
      print "\nError: Current system (%s) isn't supported\n" % host

   LookupError: unknown encoding: string_escape:
    File "/home/petr/testing/openwrt/build_dir/target-mips_24kc_musl/iotivity-1.3.1/SConstruct", line 28:
      SConscript('build_common/SConscript')

 * packages/net/smartsnmpd (KO, seems dead, no commit since 2015):

   SyntaxError: Missing parentheses in call to 'print'
    File "/openwrt/build_dir/target-mips_24kc_musl/smartsnmpd-2014-08-13/SConstruct", line 156
      print "Can't find liblua or liblua5.1!"

 * packages/utils/gpsd (KO, doesn't build even with latest 3.18.1 release):

   AttributeError: 'list' object has no attribute 'keys':
     File "/openwrt/build_dir/target-mips_24kc_musl/gpsd-3.18.1/SConstruct", line 1758:
       all_manpages = list(base_manpages.keys()) + list(python_manpages.keys())

Signed-off-by: Petr Štetiar <ynezz@true.cz>

diizzyy commented on include/prereq-build.mk:
```diff
@@ -141,10 +141,11 @@ $(eval $(call SetupHostCommand,wget,Please install GNU 'wget', \
+$(eval $(call SetupHostCommand,python,Please install Python 3.x, \
+	python3 -V 2>&1 | grep 'Python 3'))
+
+$(eval $(call SetupHostCommand,python3,Please install Python 3.x, \
```
Why two separate entries?

ynezz replied:
To get both `python` and `python3` symlinks pointing to Python 3 binary:

```
staging_dir/host/bin/python -> /usr/bin/python3
staging_dir/host/bin/python3 -> /usr/bin/python3
```

without that I wasn't able to force use of Python 3 on hosts with Python 2 and Python 3 installed, where it's usually Python 2 installed as /usr/bin/python so all build packages searching for Python would use Python 2 version instead of Python 3.

diizzyy commented:
What version of python did you test this on?
Looking at https://www.python.org/dev/peps/pep-0478/ this seems to be the oldest supported version as of writing which should be the bare minimum?

ynezz commented:
> What version of python did you test this on?

Debian GNU/Linux 9.8 (stretch), Python 3.5.3.

> which should be the bare minimum?

Can you provide one valid reason why we should enforce this right now?

diizzyy commented:
Because it makes sense to limit it to something that's verified working instead of blindly assuming that it should run on anything unless someone is willing to try previous released versions and keep supporting those in the long run.

ynezz commented:
FYI `Ubuntu 14.04.6 LTS` is still on `Python 3.4.3`.

diizzyy commented:
It's also EOL within 2 weeks
https://wiki.ubuntu.com/Releases

diizzyy commented:
To clarify, the goal isn't to elimininate users instead it's to create a "least effort" tradeoff, improve upstream and "offload" support to some extent.

If a supported version is used it's easy to report and usually receive upstream support and thus get a fix in upstream if needed. Relying on old unsupported versions will just give you more work and make you maintain your own set of patches which will take even more time to maintain as code diverges even further when someone decides that X needs to get bumped. (The short story of https://ftp.heanet.ie/mirrors/fosdem-video/2019/Janson/netflix_freebsd.mp4 @ 23:49)

ynezz commented:
I've just found two valid reasons to demand support for Python 3.5+

* meson - https://github.com/mesonbuild/meson/blob/master/setup.py#L21
* scons - https://github.com/SCons/scons/blob/master/src/script/scons.py#L64

diizzyy commented:
Also, is it a bad idea to make the shebang line consistent? :-)

ynezz committed:
build: cleanup possibly dangling Python 2 host symlink

When bumping buildroot to Python 3, we need to assure, that Python
symlink in staging bin directory points to Python >= 3.5 as well.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz commented:
> Also, is it a bad idea to make the shebang line consistent? :-)

Should be done.

ynezz commented:
@yangbolu1991 while fixing the remaining python2 bits in the tree, I've [tried](https://github.com/openwrt/openwrt/pull/1937/commits/0f31e1620735c704cf771f8244d4f3069164f256) to make `ls-rcw` package working with python3, but fixing `rcw.py` is above my head. Do you think, that this could be handled by upstream?

hauke commented on include/prereq-build.mk:
```diff
@@ -141,10 +141,19 @@ $(eval $(call SetupHostCommand,wget,Please install GNU 'wget', \
+	python3.7 -V 2>&1 | grep -E 'Python 3\.[5-9]\.?', \
+	python3.6 -V 2>&1 | grep -E 'Python 3\.[5-9]\.?', \
+	python3.5 -V 2>&1 | grep -E 'Python 3\.[5-9]\.?', \
+	python3 -V 2>&1 | grep -E 'Python 3\.[5-9]\.?'))
```
This could cause problems when python 3.10 is used in the future.

diizzyy replied:
I think we can worry about it when/if that time comes

hauke reviewed (commented):
Nice work.

I think we can remove iotivity if it causes problems, I am not using it any more.

yangbolu1991 commented:
@ynezz 
Hi Petr,
Thanks for your patches. I had asked our RCW owner for reviewing the changes.
I will update the feedback soon.
Thanks!

ja-pa commented:
@yangbolu1991 Any update?

yangbolu1991 commented:
@ynezz @ja-pa Sorry for my late response.
I haven't got reviewing feedback from our rcw owner. But they are considering to convert to python3 too. I will also confirm when will they plan to complete python3 support.

BTW, I tried to build with the pull request, but it seemed to be some build issues.
The rcw changes of the pull request is ok for me if building and the binaries work. (I can help to verify.) 

make[5]: Entering directory '/opt/samba/nxa07204/project/openwrt/upstream/openwrt/build_dir/target-aarch64_generic_musl/ls-rcw-lsdk-1806/t1042d4rdb'
python ../rcw.py -i RR_P_86/rcw_sben_1400MHz_single_source.rcw -o RR_P_86/rcw_sben_1400MHz_single_source.bin
Traceback (most recent call last):
  File "../rcw.py", line 1012, in <module>
    parse_source_file(source)
  File "../rcw.py", line 424, in parse_source_file
    pbi += parse_subsection(header, s)
  File "../rcw.py", line 393, in parse_subsection
    return build_pbi(lines)
  File "../rcw.py", line 289, in build_pbi
    m = re.match(r'\s*([a-z]+)(|\.b1|\.b2|\.b4|\.short|\.long)\s*(?<=\s)([^,]*),?([^,]*),?([^,]*),?([^,]*)', l + ' ')
TypeError: can't concat bytes to str
../Makefile.inc:17: recipe for target 'RR_P_86/rcw_sben_1400MHz_single_source.bin' failed




Thanks a lot.

zacharyrs committed:
wireless-regdb: update to 2019.06.03

Fixes build issues on a python3 host (issues with the print statement
formatting in the current build).

Includes 100-regdb-write-firmware-file-format-version-code-20.patch and
other fixes.

Closes bugs.openwrt.org/index.php?do=details&task_id=1605.

Uses the tarball as requested.

Signed-off-by: Zachary Riedlshah <git@zacharyrs.me>

ynezz commented:
Going to send patches for the review on the mailing list.

ynezz commented:
> I haven't got reviewing feedback from our rcw owner. But they are considering to convert to python3 too. I will also confirm when will they plan to complete python3 support.

@yangbolu1991 can you please check again? I'm just finishing build testing of the python3 changes and I'm going to post them for the final review on the mailing list probably tomorrow. So far it seems like `ls-rcw-lsdk-1903` is the only package having issues with python3 and would need to be fixed.

yangbolu1991 commented:
@ynezz Our rcw owner didn't have plan to support python3. Let me do it. I will try to create a pull request so that you can review and cherry-pick to use.
Thanks a lot.

yangbolu1991 commented:
See https://github.com/openwrt/openwrt/pull/2149
